components/archiving/projectController.js

```diff
@@ -0,0 +1,436 @@
+
+/**
+ * Request all project archives.
+ * Can optionally filter for items that are "not" at a certain state of archiving: 'copied-to-staging', 'sent-offsite', 'sent-to-vidarr-archival', 'unloaded'
+ */
+const allProjectArchives = async (req, res, next) => {
+  try {
+    const query = req.query.not;
+    let projects;
+    if (query) {
+      if (query == urls.filesCopiedToOffsiteStagingDir) {
+        projects = await archiveDao.getByFilesNotCopiedToOffsiteStagingDir(
+          projectEntityType
+        );
+        res.status(200).send(replaceGenericKeyName(projects));
+      } else if (query == urls.filesSentOffsite) {
+        projects = await archiveDao.getByFilesNotSentOffsite(projectEntityType);
+        res.status(200).send(replaceGenericKeyName(projects));
+      } else if (query == urls.filesLoadedIntoVidarrArchival) {
+        projects = await archiveDao.getByFilesNotLoadedIntoVidarrArchival(
+          projectEntityType
+        );
+        res.status(200).send(replaceGenericKeyName(projects));
+      } else if (query == urls.filesUnloaded) {
+        projects = await archiveDao.getByFilesNotUnloaded(projectEntityType);
+        res.status(200).send(replaceGenericKeyName(projects));
+      }
+    } else {
+      projects = await archiveDao.streamAllProjects((stream) => {
+        res.status(200);
+        stream
+          .pipe(createProjectKeyReplacementStream())
+          .pipe(JSONStream.stringify())
+          .pipe(res);
+        stream.on('error', (err) => {
+          // log the error and prematurely end the response
+          logger.error(err);
+          res.end();
+        });
+      });
+      logger.info({
+        streamRowsProcessedProjects: projects.processed,
+        streamingDurationProjects: projects.duration,
+        method: '',
+      });
+    }
+    next();
+  } catch (e) {
+    handleErrors(e, 'Error getting projects', logger, next);
+  }
+};
+
+const isCompletelyArchived = (kase) => {
+  return (
+    kase.filesCopiedToOffsiteArchiveStagingDir != null &&
+    kase.commvaultBackupJobId != null &&
+    kase.filesLoadedIntoVidarrArchival != null &&
+    kase.filesUnloaded != null
+  );
+};
+
+const hasArchivingStarted = (kase) => {
+  return kase.filesCopiedToOffsiteArchiveStagingDir != null;
+};
+
+const addProjectArchive = async (req, res, next) => {
+  try {
+    //authenticate api-key from header before continuing
+    await authenticator.authenticateRequest(req);
+
+    const existingProjects = await archiveDao.getByArchiveEntityIdentifier(
+      req.body.projectIdentifier,
+      false,
+      projectEntityType
+    );
+    if (existingProjects == null || !existingProjects.length) {
+      await upsert(req.body, true);
+      res.status(201).end();
+    } else {
+      for (let existingProject of existingProjects) {
+        // check for conflicting changes
+        let errors = [];
+        let projectStoppedProcessing = false;
+        if (existingProject.requisitionId != req.body.requisitionId) {
```

Comment:
requisition should probably be null for all project archives, as one project will be associated with multiple requisitions through its samples